File: FluidEngineIOS/TextRendering/MBETextMesh.m
//
//  MBETextMesh.m
//  TextRendering
//
//  Created by Warren Moore on 2/9/15.
//  Copyright (c) 2015 Metal By Example. All rights reserved.
//
// modified by Sebastian Detering to update the text meshes. on 10/19/2022

#import "MBETextMesh.h"
#import "MBETypes.h"
@import CoreText;

typedef void (^MBEGlyphPositionEnumerationBlock)(CGGlyph glyph,
                                                 NSInteger glyphIndex,
                                                 CGRect glyphBounds);

@implementation MBETextMesh

@synthesize vertexBuffer=_vertexBuffer;
@synthesize indexBuffer=_indexBuffer;

@synthesize rect=_rect;
@synthesize fontAtlasRef=_fontAtlasRef;
@synthesize size=_size;
@synthesize device=_device;
@synthesize justifyCenter=_justifyCenter;

- (instancetype)initWithString:(NSString *)string
                        inRect:(CGRect)rect
                      withFontAtlas:(MBEFontAtlas *)fontAtlas
                        atSize:(CGFloat)fontSize
                        device:(id<MTLDevice>)device
//                 justifyCenter:(bool)justifyCenter
{
    if ((self = [super init]))
    {
        [self buildMeshWithString:string inRect:rect withFont:fontAtlas atSize:fontSize device:device xCentering:0.0 yCentering:0.0 firstTime:true];
    }
    return self;
}

- (void)buildMeshWithString:(NSString *)string
                     inRect:(CGRect)rect
                   withFont:(MBEFontAtlas *)fontAtlas
                     atSize:(CGFloat)fontSize
                     device:(id<MTLDevice>)device
                 xCentering:(float)xCentering
                 yCentering:(float)yCentering
                firstTime:(bool)firstTime
//              justifyCenter:(bool)justifyCenter
{
    const bool shouldCenter = true;
    _rect = rect;
    _fontAtlasRef = fontAtlas;
    _size = fontSize;
    _device = device;
    UIFont *font = [fontAtlas.parentFont fontWithSize:fontSize];
    NSDictionary *attributes = @{ NSFontAttributeName : font };
    NSAttributedString *attrString = [[NSAttributedString alloc] initWithString:string attributes:attributes];
    CFRange stringRange = CFRangeMake(0, attrString.length);
    CGPathRef rectPath = CGPathCreateWithRect(rect, NULL);
    CTFramesetterRef framesetter = CTFramesetterCreateWithAttributedString((__bridge CFAttributedStringRef)attrString);
    CTFrameRef frame = CTFramesetterCreateFrame(framesetter, stringRange, rectPath, NULL);

    __block CFIndex frameGlyphCount = 0;
    NSArray *lines = (__bridge id)CTFrameGetLines(frame);
    [lines enumerateObjectsUsingBlock:^(id lineObject, NSUInteger lineIndex, BOOL *stop) {
        frameGlyphCount += CTLineGetGlyphCount((__bridge CTLineRef)lineObject);
    }];

    const size_t vertexCount = frameGlyphCount * 4;
    const size_t indexCount = frameGlyphCount * 6;
    MBEVertex *vertices = malloc(vertexCount * sizeof(MBEVertex));
    MBEIndexType *indices = malloc(indexCount * sizeof(MBEIndexType));

    __block MBEIndexType v = 0, i = 0;
    // centering variables
    __block float totalXMax = 0.0;
    __block float totalYMax = 0.0;
    [self enumerateGlyphsInFrame:frame block:^(CGGlyph glyph, NSInteger glyphIndex, CGRect glyphBounds) {
        if (glyph >= fontAtlas.glyphDescriptors.count)
        {
            NSLog(@"Font atlas has no entry corresponding to glyph #%d; Skipping...", glyph);
            return;
        }
        MBEGlyphDescriptor *glyphInfo = fontAtlas.glyphDescriptors[glyph];
        float minX = CGRectGetMinX(glyphBounds);
        float maxX = CGRectGetMaxX(glyphBounds);
        float minY = CGRectGetMinY(glyphBounds);
        float maxY = CGRectGetMaxY(glyphBounds);
        if( totalXMax < maxX ) {
            totalXMax = maxX;
        }
        if( totalYMax < maxY ) {
            totalYMax = maxY;
        }
        float minS = glyphInfo.topLeftTexCoord.x;
        float maxS = glyphInfo.bottomRightTexCoord.x;
        float minT = glyphInfo.topLeftTexCoord.y;
        float maxT = glyphInfo.bottomRightTexCoord.y;
        vertices[v++] = (MBEVertex){ { minX - xCentering, maxY - yCentering, 0, 1 }, { minS, maxT } };
        vertices[v++] = (MBEVertex){ { minX - xCentering, minY - yCentering, 0, 1 }, { minS, minT } };
        vertices[v++] = (MBEVertex){ { maxX - xCentering, minY - yCentering, 0, 1 }, { maxS, minT } };
        vertices[v++] = (MBEVertex){ { maxX - xCentering, maxY - yCentering, 0, 1 }, { maxS, maxT } };
        indices[i++] = glyphIndex * 4;
        indices[i++] = glyphIndex * 4 + 1;
        indices[i++] = glyphIndex * 4 + 2;
        indices[i++] = glyphIndex * 4 + 2;
        indices[i++] = glyphIndex * 4 + 3;
        indices[i++] = glyphIndex * 4;
    }];

    totalXMax = totalXMax / 2;
    totalYMax = totalYMax / 2;
    if( shouldCenter && firstTime ){
        [ self buildMeshWithString:string inRect:rect withFont:fontAtlas atSize:fontSize device:device xCentering:totalXMax yCentering:totalYMax firstTime:false ];
        return;
    }
    
    _vertexBuffer = [device newBufferWithBytes:vertices
                                        length:vertexCount * sizeof(MBEVertex)
                                       options:MTLResourceOptionCPUCacheModeDefault];
    [_vertexBuffer setLabel:@"Text Mesh Vertices"];
    _indexBuffer = [device newBufferWithBytes:indices
                                       length:indexCount * sizeof(MBEIndexType)
                                      options:MTLResourceOptionCPUCacheModeDefault];
    [_indexBuffer setLabel:@"Text Mesh Indices"];

    free(indices);
    free(vertices);
    CFRelease(frame);
    CFRelease(framesetter);
    CFRelease(rectPath);
}

- (void)enumerateGlyphsInFrame:(CTFrameRef)frame
                         block:(MBEGlyphPositionEnumerationBlock)block
{
    if (!block)
        return;

    CFRange entire = CFRangeMake(0, 0);

    CGPathRef framePath = CTFrameGetPath(frame);
    CGRect frameBoundingRect = CGPathGetPathBoundingBox(framePath);

    NSArray *lines = (__bridge id)CTFrameGetLines(frame);

    CGPoint *lineOriginBuffer = malloc(lines.count * sizeof(CGPoint));
    CTFrameGetLineOrigins(frame, entire, lineOriginBuffer);

    __block CFIndex glyphIndexInFrame = 0;

    UIGraphicsBeginImageContext(CGSizeMake(1, 1));
    CGContextRef context = UIGraphicsGetCurrentContext();

    [lines enumerateObjectsUsingBlock:^(id lineObject, NSUInteger lineIndex, BOOL *stop) {
        CTLineRef line = (__bridge CTLineRef)lineObject;
        CGPoint lineOrigin = lineOriginBuffer[lineIndex];

        NSArray *runs = (__bridge id)CTLineGetGlyphRuns(line);
        [runs enumerateObjectsUsingBlock:^(id runObject, NSUInteger rangeIndex, BOOL *stop) {
            CTRunRef run = (__bridge CTRunRef)runObject;

            size_t glyphCount = CTRunGetGlyphCount(run);

            CGGlyph *glyphBuffer = malloc(glyphCount * sizeof(CGGlyph));
            CTRunGetGlyphs(run, entire, glyphBuffer);

            CGPoint *positionBuffer = malloc(glyphCount * sizeof(CGPoint));
            CTRunGetPositions(run, entire, positionBuffer);

            for (size_t glyphIndex = 0; glyphIndex < glyphCount; ++glyphIndex)
            {
                CGGlyph glyph = glyphBuffer[glyphIndex];
                CGPoint glyphOrigin = positionBuffer[glyphIndex];
                CGRect glyphRect = CTRunGetImageBounds(run, context, CFRangeMake(glyphIndex, 1));
                CGFloat boundsTransX = frameBoundingRect.origin.x + lineOrigin.x;
                CGFloat boundsTransY = CGRectGetHeight(frameBoundingRect) + frameBoundingRect.origin.y - lineOrigin.y + glyphOrigin.y;
                CGAffineTransform pathTransform = CGAffineTransformMake(1, 0, 0, -1, boundsTransX, boundsTransY);
                glyphRect = CGRectApplyAffineTransform(glyphRect, pathTransform);
                block(glyph, glyphIndexInFrame, glyphRect);

                ++glyphIndexInFrame;
            }

            free(positionBuffer);
            free(glyphBuffer);
        }];
    }];
    
    UIGraphicsEndImageContext();
}

- (void) updateTextMeshWithString:(NSString *)text {
    [self buildMeshWithString:text inRect:_rect withFont:_fontAtlasRef atSize:_size device:_device xCentering:0.0 yCentering:0.0 firstTime:true] ;
}

@end

File: FluidEngineIOS/MBEText/MBETextMesh.swift
import MetalKit

class MBETextMesh: Mesh {
    
    private var _vertexBuffer: MTLBuffer!
    private var _indexBuffer: MTLBuffer!
    private var _atlasTexture: Data!
    
    private var _vertexCount: Int!
    private var _indexCount: Int!

    
    init(_ withString: String, inRect:CGRect, fontAtlas: MBEFontAtlas, atSize: CGFloat) {
        super.init()
        buildMeshWithString(string: withString, inRect: inRect, fontAtlas: fontAtlas, atSize: atSize)
        _atlasTexture = fontAtlas.textureData
    }
    
    func buildMeshWithString(string: String, inRect: CGRect, fontAtlas: MBEFontAtlas, atSize: CGFloat) {
        guard let font = fontAtlas.parentFont else { fatalError("buildMeshWithString ERROR::font from fontAtlas was nil")}
        let attributes = [NSAttributedString.Key.font: font]
        let attrString = NSAttributedString.init(string: string, attributes: attributes)
        let stringRange = CFRangeMake( 0, attrString.length )
        let rectPath = CGPath(rect: inRect, transform: nil)
        let framesetter = CTFramesetterCreateWithAttributedString(attrString)
        let frame = CTFramesetterCreateFrame(framesetter, stringRange, rectPath, nil)
        var frameGlyphCount: CFIndex = 0
        let lines = CTFrameGetLines(frame)
        (lines as NSArray).enumerateObjects { lineObject, lineIndex, stopPtr in
            frameGlyphCount += CTLineGetGlyphCount(lineObject as! CTLine)
        }
        
        let vertexCount = frameGlyphCount * 4
        let indexCount = frameGlyphCount * 6
        _indexCount = indexCount
        _vertexCount = vertexCount
        
        var vertices: [MBEVertex] = [MBEVertex].init(repeating: MBEVertex(), count: vertexCount)
        var indices: [MBEIndexType] = [MBEIndexType].init(repeating: 0, count: indexCount)
        
        var v: MBEIndexType = 0
        var i: MBEIndexType = 0
        
        enumerateGlyphs(frame) { glyph, glyphIndex, glyphBounds in
            if( glyph >= fontAtlas.glyphDescriptors.count) {
                print("MBETextMesh.buildMeshWithString() WARNING::Font atlas has no entry corresponding to glyph \(glyph)")
                return
            }
            
            let glyphInfo = fontAtlas.glyphDescriptors[Int(glyph)]
            let minX = Float(glyphBounds.minX)
            let maxX = Float(glyphBounds.maxX)
            let minY = Float(glyphBounds.minY)
            let maxY = Float(glyphBounds.maxY)
            let minS = Float(glyphInfo.topLeftTexCoord.x)
            let maxS = Float(glyphInfo.bottomRightTexCoord.x)
            let minT = Float(glyphInfo.topLeftTexCoord.y)
            let maxT = Float(glyphInfo.bottomRightTexCoord.y)
            vertices[Int(v.advanced(by: 1))] = MBEVertex(position: packed_float4(minX, maxY,0,1), texCoords: packed_float2(minS, maxT))
            vertices[Int(v.advanced(by: 1))] = MBEVertex(position: packed_float4(minX, minY, 0, 1), texCoords: packed_float2(minS, minT))
            vertices[Int(v.advanced(by: 1))] = MBEVertex(position: packed_float4(maxX, minY, 0, 1), texCoords: packed_float2(maxS, minT))
            vertices[Int(v.advanced(by: 1))] = MBEVertex(position: packed_float4(maxX, minY, 0, 1), texCoords: packed_float2(maxS, maxT) )
            indices[Int(i.advanced(by: 1))] = MBEIndexType(glyphIndex) * 4
            indices[Int(i.advanced(by: 1))] = MBEIndexType(glyphIndex) * 4 + 1
            indices[Int(i.advanced(by: 1))] = MBEIndexType(glyphIndex) * 4 + 2
            indices[Int(i.advanced(by: 1))] = MBEIndexType(glyphIndex) * 4 + 2
            indices[Int(i.advanced(by: 1))] = MBEIndexType(glyphIndex) * 4 + 3
            indices[Int(i.advanced(by: 1))] = MBEIndexType(glyphIndex) * 4
        }
        
        _vertexBuffer = Engine.Device.makeBuffer(bytes: &vertices, length: MBEVertex.size( vertexCount) )
        _vertexBuffer.label = "Text Mesh Vertices"
        _indexBuffer = Engine.Device.makeBuffer(bytes: &indices, length: MBEIndexType.size(indexCount))
        _indexBuffer.label = "Text Mesh Indices"
    }
    func enumerateGlyphs(_ inFrame: CTFrame,
                    closure: ( CGGlyph, Int, CGRect ) -> Void ){
        
        let entire: CFRange = CFRangeMake(0, 0)
        let framePath = CTFrameGetPath(inFrame)
        let frameBoundingRect = framePath.boundingBoxOfPath
        
        let lines: NSArray = CTFrameGetLines(inFrame)
        let lineOriginBuffer = [CGPoint].init(repeating: CGPoint(x:0,y:0), count: lines.count)
        var glyphIndexInFrame: CFIndex = 0
        UIGraphicsBeginImageContext(CGSize(width:1,height:1))
        
        let context = UIGraphicsGetCurrentContext()
        
        lines.enumerateObjects { lineObject, lineIndex, stopPtr in
            let line: CTLine = lineObject as! CTLine
            let lineOrigin = lineOriginBuffer[lineIndex]
            
            let runs: NSArray = CTLineGetGlyphRuns(line)
            runs.enumerateObjects { runObject, rangeIndex, stopPtr in
                let run = runObject as! CTRun
                
                let glyphCount = CTRunGetGlyphCount(run)
                
                var glyphBuffer = [CGGlyph].init(repeating: 0, count: glyphCount)
                
                CTRunGetGlyphs(run, entire, &glyphBuffer)
                
                var positionBuffer = [CGPoint].init(repeating: CGPoint(x:0,y:0), count: glyphCount)
                CTRunGetPositions(run, entire, &positionBuffer)
                
                for glyphIndex in 0..<glyphCount {
                    let glyph = glyphBuffer[glyphIndex]
                    let glyphOrigin = positionBuffer[glyphIndex]
                    var glyphRect = CTRunGetImageBounds(run, context, CFRangeMake(glyphIndex, 1))
                    let boundsTransX = frameBoundingRect.origin.x + lineOrigin.x
                    let boundsTransY = frameBoundingRect.height + frameBoundingRect.origin.y - lineOrigin.y + glyphOrigin.y
                    let pathTransform: CGAffineTransform = __CGAffineTransformMake(1, 0, 0, -1, boundsTransX, boundsTransY)
                    glyphRect = glyphRect.applying(pathTransform)
                    closure( glyph, glyphIndexInFrame, glyphRect )
                    
                    glyphIndexInFrame += 1
                }
            }
        }
        UIGraphicsEndImageContext()
    }
    
}

extension MBETextMesh: Renderable {
    func doRender(_ renderCommandEncoder: MTLRenderCommandEncoder) {
        renderCommandEncoder.setVertexBytes(&_atlasTexture, length: _atlasTexture.count, index: 2)
        renderCommandEncoder.setVertexBytes(&_vertexBuffer, length: _vertexCount, index: 3)
        renderCommandEncoder.drawIndexedPrimitives(type: .triangle,
                                                   indexCount: _indexCount,
                                                   indexType: .uint16, indexBuffer: _indexBuffer,
                                                   indexBufferOffset: 0)
        
    }
}

// MARK: refactor into proper places
//struct MBEVertex {
//    var position: packed_float4 = packed_float4(0)
//    var texCoords: packed_float2 = packed_float2(0)
//}
extension MBEVertex : sizeable { }

typealias MBEIndexType = UInt16
extension MBEIndexType: sizeable { }
